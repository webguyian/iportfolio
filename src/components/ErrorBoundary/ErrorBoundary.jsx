import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Button from 'components/Button/Button';
import Text from 'components/Text/Text';

import { isIE } from 'modules/browser/helpers';

class ErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      hasError: false,
      message: 'Shake vigorously to refresh.'
    };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidMount() {
    window.addEventListener('devicemotion', this.detectMotion, false);

    if (isIE()) {
      this.setState({
        hasError: true,
        message: "You're using an outdated and unsupported browser."
      });
    }
  }

  refresh() {
    window.location.reload();
  }

  render() {
    const { children } = this.props;
    const { hasError, message } = this.state;
    const baseClass = 'ui-error';

    if (hasError) {
      return (
        <div className={baseClass}>
          <Button
            className={`${baseClass}-btn`}
            icon="skull-crossbones"
            size="3x"
            onClick={this.refresh}
          >
            Click to refresh
          </Button>
          <Text className={`${baseClass}-heading`} element="h1" type="display">
            Something went terribly wrong.
          </Text>
          <Text className={`${baseClass}-message`} element="p">
            {message}
          </Text>
        </div>
      );
    }

    return children;
  }
}

export default ErrorBoundary;
