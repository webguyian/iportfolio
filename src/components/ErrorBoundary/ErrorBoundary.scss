.ui-error {
  @include stretch-column;
  justify-content: center;
  background-color: $blue-bright;
  text-align: center;

  &-btn {
    color: $white;
  }

  &-heading {
    margin: $gutter 0;
  }
}
