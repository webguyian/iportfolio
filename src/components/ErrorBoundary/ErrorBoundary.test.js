import React from 'react';
import { render } from '@testing-library/react';

import ErrorBoundary from './ErrorBoundary';

describe('<ErrorBoundary />', () => {
  console.error = jest.fn();

  /* eslint-disable */
  const Component = ({ text }) => (
    <div>Component goes here {text.undefined}</div>
  );
  /* eslint-enable */
  const props = {
    children: <Component />
  };

  it('renders correctly with error', () => {
    const { asFragment } = render(<ErrorBoundary {...props} />);

    expect(asFragment()).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });

  it('renders correctly without error', () => {
    const { asFragment } = render(
      <ErrorBoundary>
        <div>Component goes here</div>
      </ErrorBoundary>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
