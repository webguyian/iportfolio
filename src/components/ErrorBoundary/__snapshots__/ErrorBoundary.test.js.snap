// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> renders correctly with error 1`] = `
<DocumentFragment>
  <div
    class="ui-error"
  >
    <button
      class="ui-btn ui-btn--with-icon ui-error-btn"
      type="button"
    >
      <i
        class="ui-icon ui-icon-skull-crossbones fas fa-skull-crossbones fa-3x"
      />
      <span
        class="ui-text ui-text--accessible"
      >
        Click to refresh
      </span>
    </button>
    <h1
      class="ui-text ui-text--display ui-error-heading"
    >
      Something went terribly wrong.
    </h1>
    <p
      class="ui-text ui-text--body ui-error-message"
    >
      Shake vigorously to refresh.
    </p>
  </div>
</DocumentFragment>
`;

exports[`<ErrorBoundary /> renders correctly without error 1`] = `
<DocumentFragment>
  <div>
    Component goes here
  </div>
</DocumentFragment>
`;
